README.rst
txsimplegeo.shared
==================

Please see our on-line docs for detailed instructions, examples, support forums, etc.:

http://simplegeo.com/docs/clients-code-libraries/python


See also the blocking (non-Twisted) versions of these libraries:

https://github.com/simplegeo/python-simplegeo-shared

http://pypi.python.org/pypi/simplegeo-shared


debian/changelog
python-txsimplegeo.shared (0.9.83) maverick; urgency=low

  * UNRELEASED

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Tue, 13 Dec 2011 17:13:05 +0000

python-txsimplegeo.shared (0.9.82) maverick; urgency=low

  * UNRELEASED

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Fri, 09 Dec 2011 01:07:30 +0000

python-txsimplegeo.shared (0.9.42) lucid; urgency=low

  * UNRELEASED

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Mon, 02 May 2011 23:49:53 +0000

python-txsimplegeo.shared (0.9.41) lucid; urgency=low

  * UNRELEASED

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Fri, 15 Apr 2011 21:10:55 +0000

python-txsimplegeo.shared (0.9.40) lucid; urgency=low

  * UNRELEASED

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Tue, 05 Apr 2011 23:11:03 +0000

python-txsimplegeo.shared (0.9.39) lucid; urgency=low

  [ Zooko Ofsimplegeo ]
  * copy in the basic signing functionality from python-oauth2, along
    with tests
  * fix missing import by copying in function to_unicode

  [ SimpleGeo Nerds ]

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Thu, 17 Feb 2011 22:10:20 +0000

python-txsimplegeo.shared (0.9.0) lucid; urgency=low

  [ Zooko Ofsimplegeo ]
  * an initial version of txsimplegeo.shared which passes all tests but
    doesn't yet have tests of using oauth or any real-world manual
    testing
  * add debian/rules (copied from the txoauth project)
  * update metadata including build-time dependencies
  * version file formatted so hudson can update the build-number stamp
    without conflicting with the manually-controlled version-number
    stamp
  * parse the new split version number (manual version number and
    automatic build number)
  * see if python-setuptools-trial is spelled this way
  * stop trying to use setuptools_trial, just use trial
  * how to run trial under coverage
  * perhaps the version of coverage available on the hudson slave
    doesn't support --branch
  * ah, we need the python-twisted-bin debian package to get trial
  * hm... how about if I spell "run trial under coverage" this way...
  * I think python-coverage gets confused by the PYTHONPATH somehow...

  [ SimpleGeo Nerds ]

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Thu, 20 Jan 2011 17:53:01 +0000

python-txsimplegeo.shared (0.9) lucid; urgency=low

  [ Zooko Ofsimplegeo ]
  * initial skeletal version

 -- Zooko Ofsimplegeo <zooko@simplegeo.com>  Wed, 08 Dec 2010 03:46:54 +0000



txsimplegeo/shared/test/test_oauth.py
import unittest

from txsimplegeo.shared.oauth import escape, normalize_parameters, signing_base, sign_request, to_unicode

class ReallyEqualMixin:
    def failUnlessReallyEqual(self, a, b, msg=None):
        self.failUnlessEqual(a, b, msg=msg)
        self.failUnlessEqual(type(a), type(b), msg="a :: %r, b :: %r, %r" % (a, b, msg))

class TestFuncs(unittest.TestCase):
   def test_to_unicode(self):
        self.failUnlessRaises(TypeError, to_unicode, '\xae')

        self.failUnlessEqual(to_unicode(':-)'), u':-)')
        self.failUnlessEqual(to_unicode(u'\u00ae'), u'\u00ae')
        self.failUnlessEqual(to_unicode('\xc2\xae'), u'\u00ae')

class OauthTest(unittest.TestCase, ReallyEqualMixin):
    def test_escape(self):
        self.failUnlessReallyEqual(escape('~'), '~')
        self.failUnlessReallyEqual(escape(u'\u2766'), '%E2%9D%A6')

    def test_signing_base(self):
        sb = signing_base('GET', 'http://example.com/api/', {}, 'sekrit')
        self.failUnlessReallyEqual(sb, ('sekrit&', 'GET&http%3A%2F%2Fexample.com%2Fapi%2F&'))

    def test_normalize_parameters(self):
        np = normalize_parameters({'d': 'b', 'c': 'd'})
        self.failUnlessReallyEqual(np, 'c=d&d=b')

        np = normalize_parameters({'a': 'b', 'c': 'd'})
        self.failUnlessReallyEqual(np, 'a=b&c=d')

        np = normalize_parameters({'a': ['e', 'b'], 'c': 'd'})
        self.failUnlessReallyEqual(np, 'a=b&a=e&c=d')

        np = normalize_parameters({'a': 'b c'})
        self.failUnlessReallyEqual(np, 'a=b%20c')

        np = normalize_parameters({'a': 'b~c'})
        self.failUnlessReallyEqual(np, 'a=b~c')

    def test_sign(self):
        authheader = sign_request('abcde', 'fghijk', 'GET', 'http://example.com/api', {}, 'example')
        self.failUnlessReallyEqual(authheader, 'OAuth realm="example", oauth_signature="8b1d544bfc74ae64784dd95b0fffaf2c48a027a4", oauth_consumer_key="abcde", oauth_signature_method="HMAC-SHA1"')

        authheader = sign_request('abcde', 'fghijk', 'GET', 'http://example.com/api', {'a': ['e', 'b'], 'c': 'd'}, 'example')
        self.failUnlessReallyEqual(authheader, 'OAuth realm="example", oauth_signature="769eeb73ed355ad177cf113bc28a85a54c8c8ff6", oauth_signature_method="HMAC-SHA1", oauth_consumer_key="abcde"')


txsimplegeo/shared/test/test_client.py
from twisted.trial import unittest
from twisted.internet import defer
from twisted.web.client import Response, ResponseDone
from twisted.web.http import PotentialDataLoss

from pyutil import jsonutil as json
from txsimplegeo.shared import BodyCollector, Client, DecodeError, Feature, StringProducer, get_body

from decimal import Decimal as D

MY_OAUTH_KEY = 'MY_OAUTH_KEY'
MY_OAUTH_SECRET = 'MY_SECRET_KEY'
TESTING_LAYER = 'TESTING_LAYER'

API_VERSION = '1.0'
API_HOST = 'api.simplegeo.com'
API_PORT = 80

class DecodeErrorTest(unittest.TestCase):
    def test_repr(self):
        body = 'this is not json'
        try:
            json.loads('this is not json')
        except ValueError, le:
            e = DecodeError(body, le)
        else:
            self.fail("We were supposed to get an exception from json.loads().")

        self.failUnless("Could not decode JSON" in e.msg, repr(e.msg))
        self.failUnless('JSONDecodeError' in repr(e), repr(e))

class FakeResponse(Response):
    def __init__(self, respchunks, headers, code):
        self.respchunks = respchunks
        self.headers = headers
        self.code = code

    def deliverBody(self, consumer):
        consumer.connectionLost(ResponseDone())

class FakeSuccessResponse(FakeResponse):
    def __init__(self, respchunks, headers):
        FakeResponse.__init__(self, respchunks, headers, code=200)

    def deliverBody(self, consumer):
        for respchunk in self.respchunks:
            consumer.dataReceived(respchunk)
        consumer.connectionLost(ResponseDone())

class FakePotentialDataLossResponse(FakeResponse):
    def deliverBody(self, consumer):
        for respchunk in self.respchunks:
            consumer.dataReceived(respchunk)
        consumer.connectionLost(PotentialDataLoss())

class SomeException(Exception):
    pass

class FakeExceptionResponse(FakeResponse):
    def deliverBody(self, consumer):
        for respchunk in self.respchunks:
            consumer.dataReceived(respchunk)
        consumer.connectionLost(PotentialDataLoss())

class FakeConsumer(object):
    def write(self, body):
        self.body = body

class MockAgent(object):
    def __init__(self, fakeresp):
        self.fakeresp = fakeresp

    def request(self, method, endpoint, bodyProducer):
        self.method = method
        self.endpoint = endpoint
        self.bodyProducer = bodyProducer
        return defer.succeed(self.fakeresp)

class StringProducerTest(unittest.TestCase):
    def test_string_producer(self):
        sp = StringProducer('abc')

        sp.pauseProducing()

        fc = FakeConsumer()
        sp.startProducing(fc)

        self.failUnlessEqual(fc.body, 'abc')

        sp.stopProducing()

class BodyCollectorTest(unittest.TestCase):
    def test_body_collector_collects_body(self):
        bc = BodyCollector()
        d = bc.start()
        bc.dataReceived('a')
        bc.dataReceived('b')
        bc.connectionLost(ResponseDone())
        def _check(res):
            self.failUnless(res is bc, (res, bc))
            self.failUnless(isinstance(res.reason, ResponseDone), (res.reason, type(res.reason)))
            self.failUnless(res.bytes == 'ab', res.bytes)
        d.addCallback(_check)
        return d

    def test_body_collector_errs_on_PotentialDataLoss(self):
        bc = BodyCollector()
        d = bc.start()
        bc.dataReceived('a')
        bc.dataReceived('b')
        bc.connectionLost(PotentialDataLoss())

        d1 = self.failUnlessFailure(d, PotentialDataLoss)
        return d1

    def test_get_body_gets_body(self):
        fakeresp = FakeSuccessResponse(['a', 'b'], {})

        d = get_body(fakeresp)
        def _check(res):
            self.failUnlessEqual(res, 'ab')
        d.addCallback(_check)
        return d

    def test_get_body_errs_on_PDL(self):
        fakeresp = FakePotentialDataLossResponse(['a', 'b'], {}, 200)

        d = get_body(fakeresp)
        d2 = self.failUnlessFailure(d, PotentialDataLoss)
        return d2

    def test_get_body_errs_on_exception(self):
        fakeresp = FakeExceptionResponse(['a', 'b'], {}, 200)

        d = get_body(fakeresp)
        d2 = self.failUnlessFailure(d, Exception)
        return d2

class ClientTest(unittest.TestCase):
    def setUp(self):
        self.client = Client(MY_OAUTH_KEY, MY_OAUTH_SECRET, API_VERSION, API_HOST, API_PORT)
        self.query_lat = D('37.8016')
        self.query_lon = D('-122.4783')

    def test_wrong_endpoint(self):
        self.assertRaises(Exception, self.client._endpoint, 'wrongwrong')

    def test_missing_argument(self):
        self.assertRaises(Exception, self.client._endpoint, 'feature')

    def test_get_point_feature(self):
        mockagent = MockAgent(FakeSuccessResponse([EXAMPLE_POINT_BODY], {'status': '200', 'content-type': 'application/json', 'thingie': "just to see if you're listening"}))
        self.client.agent = mockagent

        d = self.client.get_feature("SG_4bgzicKFmP89tQFGLGZYy0_34.714646_-86.584970")
        def check_res(res):
            self.assertEqual(mockagent.method, 'GET')
            self.assertEqual(mockagent.endpoint, 'http://api.simplegeo.com:80/%s/features/%s.json' % (API_VERSION, "SG_4bgzicKFmP89tQFGLGZYy0_34.714646_-86.584970"))
            # the code under test is required to have json-decoded this before handing it back
            self.failUnless(isinstance(res, Feature), (repr(res), type(res)))
            self.failUnlessEqual(res._http_response.headers, {'status': '200', 'content-type': 'application/json', 'thingie': "just to see if you're listening"})

        d.addCallback(check_res)
        return d

    def test_get_polygon_feature(self):
        mockagent = MockAgent(FakeSuccessResponse([EXAMPLE_BODY], {'status': '200', 'content-type': 'application/json'}))
        self.client.agent = mockagent

        d = self.client.get_feature("SG_4bgzicKFmP89tQFGLGZYy0_34.714646_-86.584970")
        def check_res(res):
            self.assertEqual(mockagent.method, 'GET')
            self.assertEqual(mockagent.endpoint, 'http://api.simplegeo.com:80/%s/features/%s.json' % (API_VERSION, "SG_4bgzicKFmP89tQFGLGZYy0_34.714646_-86.584970"))
            # the code under test is required to have json-decoded this before handing it back
            self.failUnless(isinstance(res, Feature), (repr(res), type(res)))

        d.addCallback(check_res)
        return d

    def test_type_check_request(self):
        self.failUnlessRaises(TypeError, self.client._request, 'whatever', 'POST', {'bogus': "non string"})

    def test_get_feature_bad_json(self):
        mockagent = MockAgent(FakeSuccessResponse([EXAMPLE_BODY, 'some crap'], {'status': '200', 'content-type': 'application/json'}))
        self.client.agent = mockagent

        d = self.client.get_feature("SG_4bgzicKFmP89tQFGLGZYy0_34.714646_-86.584970")
        self.failUnlessFailure(d, DecodeError)
        def after_error(f):
            self.assertEqual(mockagent.endpoint, 'http://api.simplegeo.com:80/%s/features/%s.json' % (API_VERSION, "SG_4bgzicKFmP89tQFGLGZYy0_34.714646_-86.584970"))
            self.assertEqual(mockagent.method, 'GET')

        d.addCallback(after_error)

        return d

    def test_dont_json_decode_results(self):
        """ _request() is required to return a deferred that fires
        with a Response that has a deliverBody that delivers the exact
        string that the HTTP server sent to it -- no transforming it,
        such as by json-decoding. """

        mockagent = MockAgent(FakeSuccessResponse(['{ "Hello": "I am a string. \xe2\x9d\xa4" }'.decode('utf-8')], {'status': '200', 'content-type': 'application/json'}))
        self.client.agent = mockagent

        d = self.client._request("http://thing", 'POST')
        d.addCallback(get_body)
        def _with_body(body):
            self.failUnlessEqual(body, '{ "Hello": "I am a string. \xe2\x9d\xa4" }'.decode('utf-8'))
        d.addCallback(_with_body)
        return d

    def test_dont_Recordify_results(self):
        """ _request() is required to return the exact string that the HTTP
        server sent to it -- no transforming it, such as by json-decoding and
        then constructing a Record. """

        EXAMPLE_RECORD_JSONSTR=json.dumps({ 'geometry' : { 'type' : 'Point', 'coordinates' : [D('10.0'), D('11.0')] }, 'id' : 'my_id', 'type' : 'Feature', 'properties' : { 'key' : 'value'  , 'type' : 'object' } })

        mockagent = MockAgent(FakeSuccessResponse([EXAMPLE_RECORD_JSONSTR], {'status': '200', 'content-type': 'application/json'}))
        self.client.agent = mockagent

        d = self.client._request("http://thing", 'POST')
        d.addCallback(get_body)
        def check(res):
            self.failUnlessEqual(res, EXAMPLE_RECORD_JSONSTR)
        d.addCallback(check)
        return d

    def test_get_feature_error(self):
        fakeresp = FakeResponse('{"message": "help my web server is confuzzled"}', {'status': '500', 'content-type': 'application/json'}, code=500)
        mockagent = MockAgent(fakeresp)
        self.client.agent = mockagent

        d = self.client.get_feature("SG_4bgzicKFmP89tQFGLGZYy0_34.714646_-86.584970")
        self.failUnlessFailure(d, FakeResponse)
        def after_error(f):
            self.failUnlessEqual(f.code, 500)
            self.assertEqual(mockagent.endpoint, 'http://api.simplegeo.com:80/%s/features/%s.json' % (API_VERSION, "SG_4bgzicKFmP89tQFGLGZYy0_34.714646_-86.584970"))
            self.assertEqual(mockagent.method, 'GET')
        d.addCallback(after_error)
        return d

EXAMPLE_POINT_BODY="""
{"geometry":{"type":"Point","coordinates":[-105.048054,40.005274]},"type":"Feature","id":"SG_6sRJczWZHdzNj4qSeRzpzz_40.005274_-105.048054@1291669259","properties":{"province":"CO","city":"Erie","name":"CMD Colorado Inc","tags":["sandwich"],"country":"US","phone":"+1 303 664 9448","address":"305 Baron Ct","owner":"simplegeo","classifiers":[{"category":"Restaurants","type":"Food & Drink","subcategory":""}],"postcode":"80516"}}
"""

EXAMPLE_BODY="""
{"geometry":{"type":"Polygon","coordinates":[[[-86.3672637,33.4041157],[-86.3676356,33.4039745],[-86.3681259,33.40365],[-86.3685992,33.4034242],[-86.3690556,33.4031137],[-86.3695121,33.4027609],[-86.3700361,33.4024363],[-86.3705601,33.4021258],[-86.3710166,33.4018012],[-86.3715575,33.4014061],[-86.3720647,33.4008557],[-86.3724366,33.4005311],[-86.3730621,33.3998395],[-86.3733156,33.3992891],[-86.3735523,33.3987811],[-86.3737383,33.3983153],[-86.3739073,33.3978355],[-86.374144,33.3971016],[-86.3741609,33.3968758],[-86.3733494,33.3976943],[-86.3729606,33.3980189],[-86.3725211,33.3984141],[-86.3718111,33.3990069],[-86.3713378,33.399402],[-86.370949,33.3997266],[-86.3705094,33.3999948],[-86.3701206,33.4003899],[-86.3697487,33.4007287],[-86.369157,33.4012791],[-86.3687682,33.401646],[-86.3684132,33.4019847],[-86.368092,33.4023798],[-86.3676694,33.4028738],[-86.3674835,33.4033113],[-86.3672975,33.4037487],[-86.3672637,33.4041157],[-86.3672637,33.4041157]]]},"type":"Feature","properties":{"category":"Island","license":"http://creativecommons.org/licenses/by-sa/2.0/","handle":"SG_4b10i9vCyPnKAYiYBLKZN7_33.400800_-86.370802","subcategory":"","name":"Elliott Island","attribution":"(c) OpenStreetMap (http://openstreetmap.org/) and contributors CC-BY-SA (http://creativecommons.org/licenses/by-sa/2.0/)","type":"Physical Feature","abbr":""},"id":"SG_4b10i9vCyPnKAYiYBLKZN7"}
"""


txsimplegeo/shared/oauth.py
# copied from https://github.com/simplegeo/python-oauth2 and modified
# to do only what we need here

import hashlib, hmac, urllib

def to_unicode(s):
    """ Convert to unicode, raise exception with instructive error
    message if s is not unicode, ascii, or utf-8. """
    if not isinstance(s, unicode):
        if not isinstance(s, str):
            raise TypeError('You are required to pass either unicode or string here, not: %r (%s)' % (type(s), s))
        try:
            s = s.decode('utf-8')
        except UnicodeDecodeError, le:
            raise TypeError('You are required to pass either a unicode object or a utf-8 string here. You passed a Python string object which contained non-utf-8: %r. The UnicodeDecodeError that resulted from attempting to interpret it as utf-8 was: %s' % (s, le,))
    return s

def escape(s):
    """Escape a URL including any /."""
    return urllib.quote(to_unicode(s).encode('utf-8'), safe='~')

def normalize_parameters(params):
    items = []
    for k, v in params.iteritems():
        # 1.0a/9.1.1 states that kvp must be sorted by key, then by
        # value, so we unpack sequence values into multiple items for
        # sorting.
        if isinstance(v, basestring):
            items.append((k, v))
        else:
            for e in v:
                items.append((k, e))

    # Encode signature parameters per Oauth Core 1.0 protocol spec
    # draft 7, section 3.6
    # (http://tools.ietf.org/html/draft-hammer-oauth-07#section-3.6)
    # Spaces must be encoded with "%20" instead of "+"
    return urllib.urlencode(sorted(items)).replace('+', '%20').replace('%7E', '~')

def signing_base(method, url, params, secret):
    sig = [
        escape(method),
        escape(url),
        escape(normalize_parameters(params)),
    ]

    key = '%s&' % escape(secret)
    raw = '&'.join(sig)
    return key, raw

def sign_request(key, secret, method, url, params, realm):
    params['oauth_consumer_key'] = key
    params['oauth_signature_method'] = 'HMAC-SHA1'

    key, raw = signing_base(method, url, params, secret)

    params['oauth_signature'] = hmac.new(key, raw, hashlib.sha1).hexdigest()

    oauth_params = ((k, escape(to_unicode(v))) for k, v in params.iteritems() if k.startswith('oauth_'))
    header_params = ('%s="%s"' % (k, v) for k, v in oauth_params)
    params_header = ', '.join(header_params)

    auth_header = 'OAuth realm="%s"' % realm
    if params_header:
        auth_header = "%s, %s" % (auth_header, params_header)

    return auth_header


txsimplegeo/shared/__init__.py
API_VERSION = '1.0'

__all__ = []

from _version import __version__
__version__ # hush pyflakes

from zope.interface import implements

from twisted.web.client import Agent, ResponseDone
from twisted.web.iweb import IBodyProducer
from twisted.python.failure import Failure
from twisted.internet import reactor
from twisted.internet.protocol import Protocol
from twisted.internet.defer import Deferred, succeed

import copy, re
from decimal import Decimal as D

from urlparse import urljoin

from pyutil import jsonutil as json
from pyutil.assertutil import precondition

# example: http://api.simplegeo.com/1.0/feature/abcdefghijklmnopqrstuvwyz.json

def json_decode(jsonstr):
    try:
        return json.loads(jsonstr)
    except (ValueError, TypeError), le:
        raise DecodeError(jsonstr, le)

def swap(tupleab):
    return (tupleab[1], tupleab[0])

def deep_swap(struc):
    if is_numeric(struc[0]):
        assert len(struc) == 2
        assert is_numeric(struc[1])
        return swap(struc)
    return [deep_swap(sub) for sub in struc]

def deep_validate_lat_lon(struc):
    precondition(isinstance(struc, (list, tuple, set)), 'argument must be a sequence (of sequences of...) numbers')
    if is_numeric(struc[0]):
        assert len(struc) == 2
        assert is_numeric(struc[1])
        assert is_valid_lat(struc[0])
        assert is_valid_lon(struc[1])
    else:
        for sub in struc:
            deep_validate_lat_lon(sub)
    return True

SIMPLEGEOHANDLE_RSTR=r"""SG_[A-Za-z0-9]{22}(?:_-?[0-9]{1,3}(?:\.[0-9]+)?_-?[0-9]{1,3}(?:\.[0-9]+)?)?(?:@[0-9]+)?$"""
SIMPLEGEOHANDLE_R= re.compile(SIMPLEGEOHANDLE_RSTR)
def is_simplegeohandle(s):
    return isinstance(s, basestring) and SIMPLEGEOHANDLE_R.match(s)

FEATURES_URL_R=re.compile("http://(.*)/features/([A-Za-z_,-]*).json$")

def is_numeric(x):
    return isinstance(x, (int, long, float, D))

def is_valid_lat(x):
    return is_numeric(x) and (x <= 90) and (x >= -90)

def is_valid_lon(x):
    return is_numeric(x) and (x <= 180) and (x >= -180.0)

class Feature:
    def __init__(self, coordinates, geomtype='Point', simplegeohandle=None, properties=None):
        """
        The simplegeohandle and the record_id are both optional -- you
        can have one or the other or both or neither.

        A simplegeohandle is globally unique and is assigned by the
        Places service. It is returned from the Places service in the
        response to a request to add a place to the Places database
        (the add_feature method).

        The simplegeohandle is passed in as an argument to the
        constructor, named "simplegeohandle", and is stored in the
        "id" attribute of the Feature instance.

        A record_id is scoped to your particular user account and is
        chosen by you. The only use for the record_id is in case you
        call add_feature and you have already previously added that
        feature to the database -- if there is already a feature from
        your user account with the same record_id then the Places
        service will return that feature to you, along with that
        feature's simplegeohandle, instead of making a second, duplicate
        feature.

        A record_id is passed in as a value in the properties dict
        named "record_id".

        geomtype is a GeoJSON geometry type such as "Point",
        "Polygon", or "Multipolygon". coordinates is a GeoJSON
        coordinates *except* that each lat/lon pair is written in
        order lat, lon instead of the GeoJSON order of lon, at.

        When txsimplegeo.shared is constructing a Feature object from
        the result of an HTTP query to the SimpleGeo service, it will
        stash a reference to the twisted.web.client.Response object in
        the "._http_response" member variable of the Feature
        object. This could be useful for debugging, investigating the
        performance of the SimpleGeo service, etc.
        """
        precondition(simplegeohandle is None or is_simplegeohandle(simplegeohandle), "simplegeohandle is required to be None or to match the regex %s" % SIMPLEGEOHANDLE_RSTR, simplegeohandle=simplegeohandle)
        record_id = properties and properties.get('record_id') or None
        precondition(record_id is None or isinstance(record_id, basestring), "record_id is required to be None or a string.", record_id=record_id, properties=properties)
        precondition(deep_validate_lat_lon(coordinates), coordinates)

        self.id = simplegeohandle
        self.coordinates = coordinates
        self.geomtype = geomtype
        self.properties = {}
        if properties:
            self.properties.update(properties)

    @classmethod
    def from_dict(cls, data):
        """
        data is a GeoJSON standard data structure, including that the
        coordinates are in GeoJSON order (lon, lat) instead of
        SimpleGeo order (lat, lon)
        """
        assert isinstance(data, dict), (type(data), repr(data))
        feature = cls(
            simplegeohandle = data.get('id'),
            coordinates = deep_swap(data['geometry']['coordinates']),
            geomtype = data['geometry']['type'],
            properties = data.get('properties')
            )

        return feature

    def to_dict(self):
        """
        Returns a GeoJSON object, including having its coordinates in
        GeoJSON standad order (lon, lat) instead of SimpleGeo standard
        order (lat, lon).
        """
        return {
            'type': 'Feature',
            'id': self.id,
            'geometry': {
                'type': self.geomtype,
                'coordinates': deep_swap(self.coordinates)
            },
            'properties': copy.deepcopy(self.properties),
        }

    @classmethod
    def from_json(cls, jsonstr):
        return cls.from_dict(json_decode(jsonstr))

    def to_json(self):
        return json.dumps(self.to_dict())

class StringProducer(object):
    implements(IBodyProducer)
    """
    If you have a string that you want to pass as the body of your
    HTTP request, wrap it in an instance of StringProducer and pass
    that as the `body' argument of Agent.request().
    """
    def __init__(self, body):
        self.body = body
        self.length = len(body)

    def startProducing(self, consumer):
        consumer.write(self.body)
        return succeed(None)

    def pauseProducing(self):
        pass

    def stopProducing(self):
        pass

class BodyCollector(Protocol):
    """
    If you want to accumulate the body of an HTTP response until it is
    all finished and then get the result in a string, then create an
    instance of BodyCollector, call its .start() method to get a
    deferred, and pass the instance of BodyCollector as the argument
    to Response.deliverBody(). When the deferred that it gave you
    calls your callback, it will pass the string containing the HTTP
    response body.
    """
    def __init__(self):
        self.finished = Deferred()
        self.bytesl = []

    def start(self):
        return self.finished

    def dataReceived(self, bytes):
        self.bytesl.append(bytes)

    def connectionLost(self, reason):
        self.reason = reason
        self.bytes = ''.join(self.bytesl)
        if isinstance(reason, ResponseDone):
            self.finished.callback(self)
        else:
            self.finished.errback(reason)

def get_body(resp):
    """
    Takes a Response object, returns a deferred that will eventually
    fire with the response body in a string, if the response indicates
    a 200 success.
    """
    bc = BodyCollector()

    resp.deliverBody(bc)

    def _collected(bc):
        return bc.bytes

    d = bc.start()
    d.addCallback(_collected)
    return d



class Client(object):
    realm = "http://api.simplegeo.com"
    endpoints = {
        'feature': 'features/%(simplegeohandle)s.json',
    }

    def __init__(self, key, secret, api_version=API_VERSION, host="api.simplegeo.com", port=80):
        self.host = host
        self.port = port
        self.key = key
        self.secret = secret
        self.api_version = api_version
        self.uri = "http://%s:%s" % (host, port)
        self.agent = Agent(reactor)

    def get_most_recent_http_headers(self):
        """ Intended for debugging -- return the most recent HTTP
        headers which were received from the server. """
        return self.headers

    def _endpoint(self, name, **kwargs):
        """Not used directly. Finds and formats the endpoints as needed for any type of request."""
        try:
            endpoint = self.endpoints[name]
        except KeyError:
            raise Exception('No endpoint named "%s"' % name)
        try:
            endpoint = endpoint % kwargs
        except KeyError, e:
            raise TypeError('Missing required argument "%s"' % (e.args[0],))
        return urljoin(urljoin(self.uri, self.api_version + '/'), endpoint)

    def get_feature(self, simplegeohandle):
        """
        Return the GeoJSON representation of a feature.

        Return a deferred which, if the request succeeds, eventually
        fires with the Feature object. If the request fails, the
        deferred instead errbacks with the twisted.web.client.Response
        object.
        """
        precondition(is_simplegeohandle(simplegeohandle), "simplegeohandle is required to match the regex %s" % SIMPLEGEOHANDLE_RSTR, simplegeohandle=simplegeohandle)
        endpoint = self._endpoint('feature', simplegeohandle=simplegeohandle)
        d = self._request(endpoint, 'GET')
        def _handle_resp(resp):
            if (resp.code / 100) not in (2, 3):
                return Failure(resp)

            d2 = get_body(resp)
            def _handle_body(body):
                f = Feature.from_json(body)
                f._http_response = resp
                return f

            d2.addCallback(_handle_body)
            return d2
        d.addCallback(_handle_resp)
        return d

    def _request(self, endpoint, method, data=None):
        """
        Not used directly by code external to this lib. Performs the
        actual request against the API, including passing the
        credentials with oauth.  Returns deferred that eventually
        fires with a twisted.web.client.Response instance.
        """
        if data is None:
            data = ''
        elif not isinstance(data, basestring):
            raise TypeError("data is required to be None or a string or unicode, not %s" % (type(data),))

        body = StringProducer(data)

# xyZ
#         # headers = request.to_header(self.realm)
#         headers = {}
#XXX         headers['User-Agent'] = 'SimpleGeo Places Client v%s' % __version__

        d = self.agent.request(method, endpoint, bodyProducer=body)

        # def _callb(resp):
        #     self.headers = resp.header
        #     if (resp.code / 100) not in ('2', '3'):
        #         raise APIError(int(self.headers['status']), content, self.headers)
# 
#             self.k
#             self.headers, content = self.http.request(endpoint, method, body=body, headers=headers)

        return d # XXX self.headers, content


class APIError(Exception):
    """Base exception for all API errors."""

    def __init__(self, code, msg, headers, description=''):
        self.code = code
        self.msg = msg
        self.headers = headers
        self.description = description

    def __str__(self):
        return self.__repr__()

    def __repr__(self):
        return "%s (#%s) %s" % (self.msg, self.code, self.description)

class DecodeError(APIError):
    """There was a problem decoding the API's response, which was
    supposed to be encoded in JSON, but which apparently wasn't."""

    def __init__(self, body, le):
        super(DecodeError, self).__init__(None, "Could not decode JSON from server.", None, repr(le))
        self.body = body

    def __repr__(self):
        return "%s content: %s" % (self.description, self.body)
